swift/aperture/util.swift

```diff
@@ -1,26 +1,213 @@
+
+    /// Signals that cause the process to exit
+    static let exitSignals = [
+      hangup,
+      interrupt,
+      quit,
+      abort,
+      alarm,
+      termination
+    ]
+
+    let rawValue: Int32
+    init(rawValue: Int32) {
+      self.rawValue = rawValue
+    }
+  }
+
+  typealias CSignalHandler = @convention(c) (Int32) -> ()
+  typealias SignalHandler = (Signal) -> Void
+
+  private static var handlers = [Signal: [SignalHandler]]()
+
+  private static var cHandler: CSignalHandler = { rawSignal in
+    let signal = Signal(rawValue: rawSignal)
+
+    guard let signalHandlers = handlers[signal] else {
+      return
+    }
+
+    for handler in signalHandlers {
+      handler(signal)
+    }
+  }
+
+  /// Handle some signals
+  static func handle(signals: [Signal], handler: @escaping SignalHandler) {
+    for signal in signals {
+      // Since Swift has no way of running code on "struct creation", we need to initialize here…
+      if handlers[signal] == nil {
+        handlers[signal] = []
+      }
+      handlers[signal]?.append(handler)
+
+      var signalAction = sigaction(
+        __sigaction_u: unsafeBitCast(cHandler, to: __sigaction_u.self),
+        sa_mask: 0,
+        sa_flags: 0
+      )
+
+      _ = withUnsafePointer(to: &signalAction) { pointer in
+        sigaction(signal.rawValue, pointer, nil)
+      }
+    }
+  }
+
+  /// Raise a signal
+  static func raise(signal: Signal) {
+    _ = Darwin.raise(signal.rawValue)
+  }
+
+  /// Ignore a signal
+  static func ignore(signal: Signal) {
+    _ = Darwin.signal(signal.rawValue, SIG_IGN)
+  }
+
+  /// Restore default signal handling
+  static func restore(signal: Signal) {
+    _ = Darwin.signal(signal.rawValue, SIG_DFL)
   }
 }
 
-private var stderr = StandardErrorOutputStream()
+extension Array where Element == SignalHandler.Signal {
+  static let exitSignals = SignalHandler.Signal.exitSignals
+}
+// MARK: -
+
 
-func printErr(_ item: Any) {
-  print(item, to: &stderr)
+// MARK: - CLI utils
+extension FileHandle: TextOutputStream {
+  public func write(_ string: String) {
+    write(string.data(using: .utf8)!)
+  }
+}
+
+struct CLI {
+  static var standardInput = FileHandle.standardOutput
+  static var standardOutput = FileHandle.standardOutput
+  static var standardError = FileHandle.standardError
+
+  static let arguments = Array(CommandLine.arguments.dropFirst(1))
+
+  /// Called when the process exits, either normally or forced
+  static var onExit: (() -> Void)? {
+    didSet {
+      guard let exitHandler = onExit else {
+        return
+      }
+
+      var isCalled = false
+      let handler = {
+        guard !isCalled else {
+          return
+        }
+        isCalled = true
+        exitHandler()
+      }
+
+      /// I'm getting this error here:
+      /// `A C function pointer cannot be formed from a closure that captures context`
+      /// Any ideas how to solve it?
+//      atexit {
+//        handler()
+//      }
```

Comment:
@LarsJK Any idea how I can solve this? I always struggle when it comes to handling C pointer stuff.

Reply:
The only way I can think of (without modifying the c code) is to use NotificationCenter. Add an observer that calls handler and post the notification in the c callback